scripts/controlnet.py

```diff
@@ -635,6 +669,14 @@ def decode_image(img) -> np.ndarray:
                 # Add mask logic if later there is a processor that accepts mask
                 # on multiple inputs.
                 input_image = [HWC3(decode_image(img['image'])) for img in image]
+                if unit_is_ad_batch and len(image) > 0 and 'mask' in image[0] and image[0]['mask'] is not None:
+                    for idx in range(len(input_image)):
+                        while len(image[idx]['mask'].shape) < 3:
+                            image[idx]['mask'] = image[idx]['mask'][..., np.newaxis]
+                        if 'inpaint' in unit.module:
```

Comment:
Is `ad_batch` used only for ControlNet inpaint?

Reply:
iirc this piece of code only applies to ad_batch. If not I will fix

Reply:
Yes it is. Your multi-input implementation also access this. I don’t other features of CN access this part. b.t.w., your multi-input implementation is very helpful for me to remove CN hook, but I still think we need to find a way to re-implement CN in some way like forge, to make our life easier